### DISCUSSION TITLE
colossalai package no longer available on PyPI

### DISCUSSION THREAD

**Discussion by boegel** [hpcaitech/ColossalAI#2014] [Community | General]
In August 2022, we installed `colossalai` from a source tarball that was downloaded from PyPI (cfr. https://github.com/easybuilders/easybuild-easyconfigs/pull/15971), but apparently now there is no `colossalai` package anymore on PyPI.

I can find traces of ColossalAI being made available through PyPI (see https://github.com/hpcaitech/ColossalAI/pull/269 for example), but that doesn't seem to be the case anymore.

What happened? Is there a reason why the `colossalai` package was removed from PyPI?

**Comment by ver217** [comment]
Hi, we publish binary distribution here https://www.colossalai.org/download/, as PYPI doesn't allow local version.

**Comment by boegel** [comment]
@ver217 Can you clarify what you mean by "local version"?

Also, can you confirm that `colossalai` *used* to be available on PyPI, but was removed there later?

**Reply by FrankLeeeee** [reply]
Yes, perhaps we should have a installation message to say that the one from pypi is only a placeholder.

**Reply by FrankLeeeee** [reply]
The reason why we moved from public pypi to our own pip source is that Colossal-AI requires cuda extension built together with CUDA and PyTorch. If we use PyPI, the user has to build these extensions when doing `pip install` which can be time-consuming (~10 min). We want to remove such trouble, therefore, we prebuilt all these wheels with different versions of CUDA and PyTorch so that the user does not need to build on their own.

**Reply by FrankLeeeee** [reply]
According to our feedback, building extensions on the spot can be very error-prone, especially for users without knowledge on system configuration and cuda.

**Reply by boegel** [reply]
I can understand that providing pre-built wheels helps a lot of people, but is that a reason to *remove* `colossalai` from PyPI?
See https://pypi.org/project/colossalai

Especially since it was there before, you're letting the door open for people to "hijack" the `colossalai` name on PyPI, which doesn't seem wise to me.

It also looks really suspicious to have it suddenly removed from PyPI, with seemingly no public record of that, especially if the `0.1.8` tagged version on GitHub isn't exactly the same source code (see https://github.com/easybuilders/easybuild-easyconfigs/pull/16693 for more details).

**Reply by FrankLeeeee** [reply]
PyPI does not allow the same repo to be created twice, which means that others cannot create a project called colossalai. However, I think not having a release on PyPI can be kind of confusing to the user, what we plan to do is to put a placeholder distribution on PyPI which will provide messages to the user to direct them to install via our project page.

**Reply by FrankLeeeee** [reply]
Having a placeholder release should make the easybuild work?

**Reply by boegel** [reply]
A placeholder would definitely help to avoid confusion.

For EasyBuild, we will just switch to downloading the sources straight from GitHub (we strongly prefer building from source whenever possible, including dependencies, since that results in faster installations due to optimizing them for the CPUs/GPUs on which they will be used)

**Reply by FrankLeeeee** [reply]
Good to hear that. By the  way, the placehold distribution is up on PyPI.
